### ISSUE TITLE
HC2, HC3, and Heteroskedasticity-Unbiased Standard Errors

### ISSUE THREAD

**Issue by kylebutts** [lrberge/fixest#414]
Hi Laurent,

I'm writing to see if there is interest in a pull-request to add HC2, HC3, and [Heteroskedasiticity-Unbiased](https://onlinelibrary.wiley.com/doi/full/10.3982/ECTA16410) standard errors into `fixest`. HC2 and HC3 are useful for small samples (e.g. see http://datacolada.org/99); the latter is useful for high-dimensional models. For example, unbiased estimates of the variance of teacher fixed-effects you need to have a HU estimator. 

A little summary (for completeness) 

HC1 estimates $\hat{\Omega}$ as a diagonal matrix with $t^{th}$ element $\frac{n}{n - k} \hat{u}_{t}^2$. 

HC2 replaces with $\hat{u}_{t}^2 / (1-h_t)$ where $h_t$ is the $t^{th}$ diagonal element of the projection matrix.

HC3 with $\hat{u}_{t}^2 / (1-h_t)^2$.

HU uses $y_t \hat{u}_t / (1 - h_t)$. 

Therefore the main thing needed is the ability to quickly calculate the diagonal of the projection matrix. We already have the code to make the sparse model matrix (from your help in github.com/kylebutts/did2s). For low-dimensions or with just fixed-effects, it's *super fast* to get the $h_t$ vector exactly (by leveraging sparseness). With large $n$, covariates and high-dimensional fixed effects, then $h_t$ can be approximated by a Johnson-Lindenstrauss approximation (from Kline, Saggio, and Sølvsten). I have code for exact and approximate calculation of $h_t$, but I can probably go to a lower level and parallelize it for more speedup. 
 
Let me know if this would be a good fit within `fixest`; otherwise I can make a small package that just calculates the $V$ and then can be "inserted" back into the `feols` object with `summary`.

**Comment by s3alfisc** [comment]
Hi @kylebutts - not sure if I misunderstand your question, but I think that you could simply use `fixest::demean()`, which runs the basic MAP algorithm for a matrix X and a list of fixed effects f. This way, you would call `demean()` only once on the rhs of your model and get the demeaned covariates. In a second step, you could either loop over all p vectors of rademacher weights, or alternatively loop over matrices of rademacher weights. As the demeaning of each dependent variable is independent of all others, it does not matter if you demean them individually or in chunks. 

```r
library(fixest)
N <- 1000
k <- 10
X <- matrix(rnorm(N*k, 0, 1), N, k)
f <- sample(1:10, N, TRUE)

res_loop <- X
for(i in 1:k){
  res_loop[,i] <- fixest::demean(X[,i], f)
}

res_all <- fixest::demean(X, f)

all.equal(res_loop, res_all)
#[1] TRUE
```

**Comment by kylebutts** [comment]
Closed in favor of #418
